torch/distributed/algorithms/join.py

```diff
@@ -0,0 +1,197 @@
+
+    .. warning::
+        The context manager requires a ``dist.all_reduce(torch.ones(1))`` to be
+        called on every non-joined process each time before it performs its
+        collective communications in order to indicate that the process has not
+        yet joined. For example, this can occur at the beginning of the forward
+        pass.
+
+    .. warning::
+        For the ``join_hook.object`` s, the context manager requires that at
+        least one has its ``process_group`` attribute set, that at least one
+        has its ``device`` attribute set, and that all specified process groups
+        are the same. This process group and device information is used for
+        checking for non-joined processes and for notifying processes to
+        terminate if ``throw_on_early_termination`` is enabled, both of which
+        using an all-reduce.
+
+    Arguments:
+        join_hooks (List[_JoinHook]): a list of the :class:`_JoinHook` s to
+            use; the hooks are iterated over in the given order.
+
+        enable (bool): a flag enabling uneven input detection; setting to
+            ``False`` disables the context manager's functionality and should
+            only be set when the user knows the inputs will not be uneven
+            (default: ``True``).
+
+        throw_on_early_termination (bool): a flag controlling whether to raise
+            an exception upon detecting uneven inputs (default: ``False``).
+
+    """
+    def __init__(
+        self,
+        join_hooks: List[_JoinHook],
+        enable: bool = True,
+        throw_on_early_termination: bool = False,
+    ):
+        if len(join_hooks) == 0:
+            raise ValueError("The join context manager requires at least one join hook")
+        self._join_hooks = join_hooks
+        self._enable = enable
+        self._throw_on_early_termination = throw_on_early_termination
+        self._extract_dist_info()
+
+    def _extract_dist_info(self):
+        r"""
+        Extracts the process group and device information from the join hooks.
+
+        Preconditions:
+            ``self._join_hooks`` is not ``None`` and is non-empty.
+
+        Raises:
+            ValueError
+                If none of the ``join_hook.object`` s have ``process_group``
+                set; if none of the ``join_hook.object`` s have ``device``
+                set; or if there are multiple conflicting ``process_group`` s.
+
+        NOTE: If there are multiple devices specified, then the context manager
+        uses the first.
+        """
+        process_group = None
+        device = None
+        for join_hook in self._join_hooks:
+            if hasattr(join_hook.object, "process_group"):
+                if process_group is None:
+                    process_group = join_hook.object.process_group
+                elif process_group != join_hook.object.process_group:
+                    raise ValueError("Using join context manager with multiple process groups")
```

Comment:
minor todo in followup PRs: we can revive the `name` attribute in process group instances and print the name here.

Reply:
It appears that this requires adding an accessor in `distributed_c10d.py` for `_pg_names`.

Reply:
If we revive the `group_name` in `init_process_group` API, it can also be an attribute in the ProcessGroup instance.